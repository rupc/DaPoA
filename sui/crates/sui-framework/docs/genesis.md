
<a name="0x2_genesis"></a>

# Module `0x2::genesis`



-  [Constants](#@Constants_0)
-  [Function `create`](#0x2_genesis_create)


<pre><code><b>use</b> <a href="">0x1::option</a>;
<b>use</b> <a href="balance.md#0x2_balance">0x2::balance</a>;
<b>use</b> <a href="clock.md#0x2_clock">0x2::clock</a>;
<b>use</b> <a href="coin.md#0x2_coin">0x2::coin</a>;
<b>use</b> <a href="epoch_time_lock.md#0x2_epoch_time_lock">0x2::epoch_time_lock</a>;
<b>use</b> <a href="sui.md#0x2_sui">0x2::sui</a>;
<b>use</b> <a href="sui_system.md#0x2_sui_system">0x2::sui_system</a>;
<b>use</b> <a href="tx_context.md#0x2_tx_context">0x2::tx_context</a>;
<b>use</b> <a href="validator.md#0x2_validator">0x2::validator</a>;
</code></pre>



<a name="@Constants_0"></a>

## Constants


<a name="0x2_genesis_INIT_MAX_VALIDATOR_COUNT"></a>

Initial value of the upper-bound on the number of validators.


<pre><code><b>const</b> <a href="genesis.md#0x2_genesis_INIT_MAX_VALIDATOR_COUNT">INIT_MAX_VALIDATOR_COUNT</a>: u64 = 100;
</code></pre>



<a name="0x2_genesis_INIT_MIN_VALIDATOR_STAKE"></a>

Initial value of the lower-bound on the amount of stake required to become a validator.
TODO: testnet only. Needs to be changed.


<pre><code><b>const</b> <a href="genesis.md#0x2_genesis_INIT_MIN_VALIDATOR_STAKE">INIT_MIN_VALIDATOR_STAKE</a>: u64 = 1;
</code></pre>



<a name="0x2_genesis_INIT_STAKE_SUBSIDY_AMOUNT"></a>

Stake subisidy to be given out in the very first epoch. Placeholder value.


<pre><code><b>const</b> <a href="genesis.md#0x2_genesis_INIT_STAKE_SUBSIDY_AMOUNT">INIT_STAKE_SUBSIDY_AMOUNT</a>: u64 = 1000000;
</code></pre>



<a name="0x2_genesis_INIT_STAKE_SUBSIDY_FUND_BALANCE"></a>

The initial balance of the Subsidy fund in Mist (1 Billion * 10^9)


<pre><code><b>const</b> <a href="genesis.md#0x2_genesis_INIT_STAKE_SUBSIDY_FUND_BALANCE">INIT_STAKE_SUBSIDY_FUND_BALANCE</a>: u64 = 1000000000000000000;
</code></pre>



<a name="0x2_genesis_create"></a>

## Function `create`

This function will be explicitly called once at genesis.
It will create a singleton SuiSystemState object, which contains
all the information we need in the system.


<pre><code><b>fun</b> <a href="genesis.md#0x2_genesis_create">create</a>(initial_sui_custody_account_address: <b>address</b>, validator_pubkeys: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_network_pubkeys: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_worker_pubkeys: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_proof_of_possessions: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_sui_addresses: <a href="">vector</a>&lt;<b>address</b>&gt;, validator_names: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_descriptions: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_image_urls: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_project_urls: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_net_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_p2p_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_consensus_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_worker_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;, validator_gas_prices: <a href="">vector</a>&lt;u64&gt;, validator_commission_rates: <a href="">vector</a>&lt;u64&gt;, protocol_version: u64, epoch_start_timestamp_ms: u64, ctx: &<b>mut</b> <a href="tx_context.md#0x2_tx_context_TxContext">tx_context::TxContext</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>fun</b> <a href="genesis.md#0x2_genesis_create">create</a>(
    initial_sui_custody_account_address: <b>address</b>,
    validator_pubkeys: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_network_pubkeys: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_worker_pubkeys: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_proof_of_possessions: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_sui_addresses: <a href="">vector</a>&lt;<b>address</b>&gt;,
    validator_names: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_descriptions: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_image_urls: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_project_urls: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_net_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_p2p_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_consensus_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_worker_addresses: <a href="">vector</a>&lt;<a href="">vector</a>&lt;u8&gt;&gt;,
    validator_gas_prices: <a href="">vector</a>&lt;u64&gt;,
    validator_commission_rates: <a href="">vector</a>&lt;u64&gt;,
    protocol_version: u64,
    epoch_start_timestamp_ms: u64,
    ctx: &<b>mut</b> TxContext,
) {
    <b>let</b> sui_supply = <a href="sui.md#0x2_sui_new">sui::new</a>(ctx);
    <b>let</b> subsidy_fund = <a href="balance.md#0x2_balance_split">balance::split</a>(&<b>mut</b> sui_supply, <a href="genesis.md#0x2_genesis_INIT_STAKE_SUBSIDY_FUND_BALANCE">INIT_STAKE_SUBSIDY_FUND_BALANCE</a>);
    <b>let</b> storage_fund = <a href="balance.md#0x2_balance_zero">balance::zero</a>();
    <b>let</b> validators = <a href="_empty">vector::empty</a>();
    <b>let</b> count = <a href="_length">vector::length</a>(&validator_pubkeys);
    <b>assert</b>!(
        <a href="_length">vector::length</a>(&validator_sui_addresses) == count
            && <a href="_length">vector::length</a>(&validator_names) == count
            && <a href="_length">vector::length</a>(&validator_descriptions) == count
            && <a href="_length">vector::length</a>(&validator_image_urls) == count
            && <a href="_length">vector::length</a>(&validator_project_urls) == count
            && <a href="_length">vector::length</a>(&validator_net_addresses) == count
            && <a href="_length">vector::length</a>(&validator_p2p_addresses) == count
            && <a href="_length">vector::length</a>(&validator_consensus_addresses) == count
            && <a href="_length">vector::length</a>(&validator_worker_addresses) == count
            && <a href="_length">vector::length</a>(&validator_gas_prices) == count
            && <a href="_length">vector::length</a>(&validator_commission_rates) == count,
        1
    );
    <b>let</b> i = 0;
    <b>while</b> (i &lt; count) {
        <b>let</b> sui_address = *<a href="_borrow">vector::borrow</a>(&validator_sui_addresses, i);
        <b>let</b> pubkey = *<a href="_borrow">vector::borrow</a>(&validator_pubkeys, i);
        <b>let</b> network_pubkey = *<a href="_borrow">vector::borrow</a>(&validator_network_pubkeys, i);
        <b>let</b> worker_pubkey = *<a href="_borrow">vector::borrow</a>(&validator_worker_pubkeys, i);
        <b>let</b> proof_of_possession = *<a href="_borrow">vector::borrow</a>(&validator_proof_of_possessions, i);
        <b>let</b> name = *<a href="_borrow">vector::borrow</a>(&validator_names, i);
        <b>let</b> description = *<a href="_borrow">vector::borrow</a>(&validator_descriptions, i);
        <b>let</b> image_url = *<a href="_borrow">vector::borrow</a>(&validator_image_urls, i);
        <b>let</b> project_url = *<a href="_borrow">vector::borrow</a>(&validator_project_urls, i);
        <b>let</b> net_address = *<a href="_borrow">vector::borrow</a>(&validator_net_addresses, i);
        <b>let</b> p2p_address = *<a href="_borrow">vector::borrow</a>(&validator_p2p_addresses, i);
        <b>let</b> consensus_address = *<a href="_borrow">vector::borrow</a>(&validator_consensus_addresses, i);
        <b>let</b> worker_address = *<a href="_borrow">vector::borrow</a>(&validator_worker_addresses, i);
        <b>let</b> gas_price = *<a href="_borrow">vector::borrow</a>(&validator_gas_prices, i);
        <b>let</b> commission_rate = *<a href="_borrow">vector::borrow</a>(&validator_commission_rates, i);
        <a href="_push_back">vector::push_back</a>(&<b>mut</b> validators, <a href="validator.md#0x2_validator_new">validator::new</a>(
            sui_address,
            pubkey,
            network_pubkey,
            worker_pubkey,
            proof_of_possession,
            name,
            description,
            image_url,
            project_url,
            net_address,
            p2p_address,
            consensus_address,
            worker_address,
            // TODO Figure out <b>if</b> we want <b>to</b> instead initialize validators <b>with</b> 0 stake.
            // Initialize all validators <b>with</b> 1 Mist stake.
            <a href="balance.md#0x2_balance_split">balance::split</a>(&<b>mut</b> sui_supply, 1),
            <a href="_none">option::none</a>(),
            gas_price,
            commission_rate,
            0, // start operating right away at epoch 0
            ctx
        ));
        i = i + 1;
    };

    <a href="sui_system.md#0x2_sui_system_create">sui_system::create</a>(
        validators,
        subsidy_fund,
        storage_fund,
        <a href="genesis.md#0x2_genesis_INIT_MAX_VALIDATOR_COUNT">INIT_MAX_VALIDATOR_COUNT</a>,
        <a href="genesis.md#0x2_genesis_INIT_MIN_VALIDATOR_STAKE">INIT_MIN_VALIDATOR_STAKE</a>,
        <a href="genesis.md#0x2_genesis_INIT_STAKE_SUBSIDY_AMOUNT">INIT_STAKE_SUBSIDY_AMOUNT</a>,
        protocol_version,
        epoch_start_timestamp_ms,
        ctx,
    );

    <a href="clock.md#0x2_clock_create">clock::create</a>();

    // Transfer the remaining <a href="balance.md#0x2_balance">balance</a> of <a href="sui.md#0x2_sui">sui</a>'s supply <b>to</b> the initial account
    <a href="sui.md#0x2_sui_transfer">sui::transfer</a>(<a href="coin.md#0x2_coin_from_balance">coin::from_balance</a>(sui_supply, ctx), initial_sui_custody_account_address);
}
</code></pre>



</details>
