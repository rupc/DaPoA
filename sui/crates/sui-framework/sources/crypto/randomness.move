// // Copyright (c) Mysten Labs, Inc.
// // SPDX-License-Identifier: Apache-2.0

/// WARNING: This module is currently insecure as the randomness source it depends on is not yet secure
/// (work-in-progress). Please use it for testing purposes only.

/// Randomness objects can only be created, set or consumed. They cannot be created and consumed
/// in the *same* transaction since it might allow validators decide whether to create and use those
/// objects *after* seeing the randomness they depend on.
///
/// - On creation, the object contains the epoch in which it was created and a unique object id.
///
/// - After the object creation transaction is committed, anyone can retrieve the BLS signature on
///   message "randomness":epoch:id from validators (signed using the Threshold-BLS key of that
///   epoch).
///
/// - Anyone that can mutate the object can set the randomness of the object by supplying the BLS
///   signature. This operation verifies the signature and sets the value of the randomness object
///   to be the hash of the signature.
///
///   Note that there is a single signature that could pass the verification for a specific object,
///   thus, the only options the owner of the object has after retrieving the signature (and learning
///   the randomness) is to either set the randomness or leave it unset. Applications that use
///   Randomness objects must make sure they handle both options (e.g., debit the user on object
///   creation so even if the user aborts, depending on the randomness it received, the application
///   is not harmed).
///
/// - Once set, the random value can be read/consumed.
///
/// This object can be used as a shared-/owned-object.
///
module sui::randomness {
    use std::hash::sha3_256;
    use std::option::{Self, Option};
    use std::vector;
    use sui::bcs;
    use sui::object::{Self, UID, ID};
    use sui::transfer;
    use sui::tx_context::{Self, TxContext};

    /// Set is called with an invalid signature.
    const EInvalidSignature: u64 = 0;
    /// Already set object cannot be set again.
    const EAlreadySet: u64 = 1;
    /// Supplied randomness is not of the right length.
    const EInvalidRndLength: u64 = 2;

    /// All signatures are prefixed with Domain.
    const Domain: vector<u8> = b"randomness";

    /// WARNING: This object is currently insecure as the randomness source it depends on is not yet secure
    /// (work-in-progress). Please use it for testing purposes only.
    struct Randomness<phantom T> has key {
        id: UID,
        epoch: u64,
        value: Option<vector<u8>>
    }

    public fun new<T: drop>(_w: T, ctx: &mut TxContext): Randomness<T> {
        Randomness<T> {
            id: object::new(ctx),
            epoch: tx_context::epoch(ctx),
            value: option::none(),
        }
        // TODO: Front load the fee.
    }

    public fun transfer<T>(self: Randomness<T>, to: address) {
        transfer::transfer(self, to);
    }

    public fun share_object<T>(self: Randomness<T>) {
        transfer::share_object(self);
    }

    /// Owner(s) can use this function for setting the randomness.
    public fun set<T>(self: &mut Randomness<T>, sig: vector<u8>) {
        assert!(option::is_none(&self.value), EAlreadySet);
        let msg = to_bytes(&Domain, self.epoch, &object::id(self));
        assert!(native_tbls_verify_signature(self.epoch, &msg, &sig), EInvalidSignature);
        let hashed = sha3_256(sig);
        self.value = option::some(hashed);
    }

    /// Delete the object.
    public fun destroy<T>(r: Randomness<T>) {
        let Randomness { id, epoch: _, value: _ } = r;
        object::delete(id);
    }

    /// Read the epoch of the object.
    public fun epoch<T>(self: &Randomness<T>): u64 {
        self.epoch
    }

    /// Read the current value of the object.
    public fun value<T>(self: &Randomness<T>): &Option<vector<u8>> {
        &self.value
    }

    fun to_bytes(domain: &vector<u8>, epoch: u64, id: &ID): vector<u8> {
        let buffer: vector<u8> = vector::empty();
        // All elements below are of fixed sizes.
        vector::append(&mut buffer, *domain);
        vector::append(&mut buffer, bcs::to_bytes(&epoch));
        vector::append(&mut buffer, object::id_to_bytes(id));
        buffer
    }

    /// Verify signature sig on message msg using the epoch's BLS key.
    native fun native_tbls_verify_signature(epoch: u64, msg: &vector<u8>, sig: &vector<u8>): bool;

    /// Helper functions to sign on messages in tests.
    native fun native_tbls_sign(epoch: u64, msg: &vector<u8>): vector<u8>;

    #[test_only]
    public fun sign<T>(self: &Randomness<T>): vector<u8> {
        let msg = to_bytes(&Domain, self.epoch, &object::id(self));
        native_tbls_sign(self.epoch, &msg)
    }

    // Helper functions for working with the outputs of Randomness

    // Converts the first 16 bytes of rnd to a u128 number and outputs its modulo with input n.
    // Since n is u64, the output is at most 2^{-64} biased assuming rnd is uniformly random.
    public fun safe_selection(n: u64, rnd: &vector<u8>): u64 {
        assert!(vector::length(rnd) >= 16, EInvalidRndLength);
        let m: u128 = 0;
        let i = 0;
        while (i < 16) {
            m = m << 8;
            let curr_byte = *vector::borrow(rnd, i);
            m = m + (curr_byte as u128);
            i = i + 1;
        };
        let n_128 = (n as u128);
        let module_128  = m % n_128;
        let res = (module_128 as u64);
        res
    }
}
