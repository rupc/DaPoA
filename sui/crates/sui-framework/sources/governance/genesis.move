// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

module sui::genesis {
    use std::vector;

    use sui::balance;
    use sui::coin;
    use sui::clock;
    use sui::sui;
    use sui::sui_system;
    use sui::tx_context::TxContext;
    use sui::validator;
    use std::option;

    /// Initial value of the lower-bound on the amount of stake required to become a validator.
    /// TODO: testnet only. Needs to be changed.
    const INIT_MIN_VALIDATOR_STAKE: u64 = 1;

    /// Initial value of the upper-bound on the number of validators.
    const INIT_MAX_VALIDATOR_COUNT: u64 = 100;

    /// Stake subisidy to be given out in the very first epoch. Placeholder value.
    const INIT_STAKE_SUBSIDY_AMOUNT: u64 = 1000000;

    /// The initial balance of the Subsidy fund in Mist (1 Billion * 10^9)
    const INIT_STAKE_SUBSIDY_FUND_BALANCE: u64 = 1_000_000_000_000_000_000;

    /// This function will be explicitly called once at genesis.
    /// It will create a singleton SuiSystemState object, which contains
    /// all the information we need in the system.
    fun create(
        initial_sui_custody_account_address: address,
        validator_pubkeys: vector<vector<u8>>,
        validator_network_pubkeys: vector<vector<u8>>,
        validator_worker_pubkeys: vector<vector<u8>>,
        validator_proof_of_possessions: vector<vector<u8>>,
        validator_sui_addresses: vector<address>,
        validator_names: vector<vector<u8>>,
        validator_descriptions: vector<vector<u8>>,
        validator_image_urls: vector<vector<u8>>,
        validator_project_urls: vector<vector<u8>>,
        validator_net_addresses: vector<vector<u8>>,
        validator_p2p_addresses: vector<vector<u8>>,
        validator_consensus_addresses: vector<vector<u8>>,
        validator_worker_addresses: vector<vector<u8>>,
        validator_gas_prices: vector<u64>,
        validator_commission_rates: vector<u64>,
        protocol_version: u64,
        epoch_start_timestamp_ms: u64,
        ctx: &mut TxContext,
    ) {
        let sui_supply = sui::new(ctx);
        let subsidy_fund = balance::split(&mut sui_supply, INIT_STAKE_SUBSIDY_FUND_BALANCE);
        let storage_fund = balance::zero();
        let validators = vector::empty();
        let count = vector::length(&validator_pubkeys);
        assert!(
            vector::length(&validator_sui_addresses) == count
                && vector::length(&validator_names) == count
                && vector::length(&validator_descriptions) == count
                && vector::length(&validator_image_urls) == count
                && vector::length(&validator_project_urls) == count
                && vector::length(&validator_net_addresses) == count
                && vector::length(&validator_p2p_addresses) == count
                && vector::length(&validator_consensus_addresses) == count
                && vector::length(&validator_worker_addresses) == count
                && vector::length(&validator_gas_prices) == count
                && vector::length(&validator_commission_rates) == count,
            1
        );
        let i = 0;
        while (i < count) {
            let sui_address = *vector::borrow(&validator_sui_addresses, i);
            let pubkey = *vector::borrow(&validator_pubkeys, i);
            let network_pubkey = *vector::borrow(&validator_network_pubkeys, i);
            let worker_pubkey = *vector::borrow(&validator_worker_pubkeys, i);
            let proof_of_possession = *vector::borrow(&validator_proof_of_possessions, i);
            let name = *vector::borrow(&validator_names, i);
            let description = *vector::borrow(&validator_descriptions, i);
            let image_url = *vector::borrow(&validator_image_urls, i);
            let project_url = *vector::borrow(&validator_project_urls, i);
            let net_address = *vector::borrow(&validator_net_addresses, i);
            let p2p_address = *vector::borrow(&validator_p2p_addresses, i);
            let consensus_address = *vector::borrow(&validator_consensus_addresses, i);
            let worker_address = *vector::borrow(&validator_worker_addresses, i);
            let gas_price = *vector::borrow(&validator_gas_prices, i);
            let commission_rate = *vector::borrow(&validator_commission_rates, i);
            vector::push_back(&mut validators, validator::new(
                sui_address,
                pubkey,
                network_pubkey,
                worker_pubkey,
                proof_of_possession,
                name,
                description,
                image_url,
                project_url,
                net_address,
                p2p_address,
                consensus_address,
                worker_address,
                // TODO Figure out if we want to instead initialize validators with 0 stake.
                // Initialize all validators with 1 Mist stake.
                balance::split(&mut sui_supply, 1),
                option::none(),
                gas_price,
                commission_rate,
                0, // start operating right away at epoch 0
                ctx
            ));
            i = i + 1;
        };

        sui_system::create(
            validators,
            subsidy_fund,
            storage_fund,
            INIT_MAX_VALIDATOR_COUNT,
            INIT_MIN_VALIDATOR_STAKE,
            INIT_STAKE_SUBSIDY_AMOUNT,
            protocol_version,
            epoch_start_timestamp_ms,
            ctx,
        );

        clock::create();

        // Transfer the remaining balance of sui's supply to the initial account
        sui::transfer(coin::from_balance(sui_supply, ctx), initial_sui_custody_account_address);
    }
}
