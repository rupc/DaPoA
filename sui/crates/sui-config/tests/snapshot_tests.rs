// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

// This file contains tests that detect changes in Sui configs.
// If a PR breaks one or more tests here, the PR probably has a real impact
// on a production configuration file. When test failure happens, the PR should
// be marked as a breaking change and reviewers should be aware of this.
//
// Owners and operators of production configuration files can add themselves to
// .github/CODEOWNERS for the corresponding snapshot tests, so they can get notified
// of changes. PRs that modifies snapshot files should wait for reviews from
// code owners (if any) before merging.
//
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta
// 1. Run `cargo insta test --review` under `./sui-config`.
// 2. Review, accept or reject changes.

use fastcrypto::traits::KeyPair;
use insta::assert_yaml_snapshot;
use multiaddr::Multiaddr;
use rand::rngs::StdRng;
use rand::SeedableRng;
use sui_config::genesis::GenesisChainParameters;
use sui_config::ValidatorInfo;
use sui_config::{genesis::Builder, genesis_config::GenesisConfig};
use sui_protocol_config::ProtocolVersion;
use sui_types::base_types::{ObjectID, SuiAddress};
use sui_types::crypto::{
    generate_proof_of_possession, get_key_pair_from_rng, AccountKeyPair, AuthorityKeyPair,
    AuthorityPublicKeyBytes, NetworkKeyPair,
};

#[test]
#[cfg_attr(msim, ignore)]
fn genesis_config_snapshot_matches() {
    // Test creating fake SuiAddress from PublicKeyBytes.
    let keypair: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let public_key = AuthorityPublicKeyBytes::from(keypair.public());
    let fake_addr = SuiAddress::from(&public_key);

    let fake_obj_id = ObjectID::from(fake_addr);
    let mut genesis_config = GenesisConfig::for_local_testing();
    genesis_config.parameters.timestamp_ms = 0;
    for account in &mut genesis_config.accounts {
        account.address = Some(fake_addr);
        for gas_obj in &mut account.gas_objects {
            gas_obj.object_id = fake_obj_id;
        }
    }
    assert_yaml_snapshot!(genesis_config);
}

#[test]
#[cfg_attr(msim, ignore)]
fn populated_genesis_snapshot_matches() {
    let genesis_config = GenesisConfig::for_local_testing();
    let (_account_keys, objects) = genesis_config
        .generate_accounts(&mut StdRng::from_seed([0; 32]))
        .unwrap();
    let key: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let worker_key: NetworkKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let network_key: NetworkKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let account_key: AccountKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let validator = ValidatorInfo {
        name: "0".into(),
        protocol_key: key.public().into(),
        worker_key: worker_key.public().clone(),
        account_key: account_key.public().clone().into(),
        network_key: network_key.public().clone(),
        gas_price: 1,
        commission_rate: 0,
        network_address: Multiaddr::empty(),
        p2p_address: Multiaddr::empty(),
        narwhal_primary_address: Multiaddr::empty(),
        narwhal_worker_address: Multiaddr::empty(),
        description: String::new(),
        image_url: String::new(),
        project_url: String::new(),
    };
    let pop = generate_proof_of_possession(&key, account_key.public().into());

    let genesis = Builder::new()
        .add_objects(objects)
        .add_validator(validator, pop)
        .with_parameters(GenesisChainParameters {
            timestamp_ms: 10,
            protocol_version: ProtocolVersion::MAX,
            allow_insertion_of_extra_objects: true,
            initial_sui_custody_account_address: SuiAddress::default(),
        })
        .add_validator_signature(&key)
        .build();
    assert_yaml_snapshot!(genesis.validator_set());
    assert_yaml_snapshot!(genesis.sui_system_wrapper_object());
    assert_yaml_snapshot!(genesis.sui_system_object());
    assert_yaml_snapshot!(genesis.clock());
    // Serialized `genesis` is not static and cannot be snapshot tested.
}

#[test]
#[cfg_attr(msim, ignore)]
fn network_config_snapshot_matches() {
    use std::net::{IpAddr, Ipv4Addr, SocketAddr};
    use std::path::PathBuf;
    use sui_config::NetworkConfig;

    let temp_dir = tempfile::tempdir().unwrap();
    let committee_size = 7;
    let rng = StdRng::from_seed([0; 32]);
    let mut network_config = NetworkConfig::generate_with_rng(temp_dir.path(), committee_size, rng);
    // TODO: Inject static temp path and port numbers, instead of clearing them.
    for mut validator_config in &mut network_config.validator_configs {
        validator_config.db_path = PathBuf::from("/tmp/foo/");
        validator_config.network_address = Multiaddr::empty();
        let fake_socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 1);
        validator_config.json_rpc_address = fake_socket;
        validator_config.metrics_address = fake_socket;
        validator_config.p2p_config.listen_address = fake_socket;
        validator_config.p2p_config.external_address = None;
        validator_config.admin_interface_port = 8888;
        let metrics_addr: Multiaddr = "/ip4/127.0.0.1/tcp/1234".parse().unwrap();
        let primary_network_admin_server_port = 5678;
        let worker_network_admin_server_base_port = 8765;
        if let Some(consensus_config) = validator_config.consensus_config.as_mut() {
            consensus_config.address = Multiaddr::empty();
            consensus_config.db_path = PathBuf::from("/tmp/foo/");
            consensus_config.internal_worker_address = Some(Multiaddr::empty());
            consensus_config
                .narwhal_config
                .consensus_api_grpc
                .socket_addr = Multiaddr::empty();
            consensus_config
                .narwhal_config
                .prometheus_metrics
                .socket_addr = metrics_addr;
            consensus_config
                .narwhal_config
                .network_admin_server
                .primary_network_admin_server_port = primary_network_admin_server_port;
            consensus_config
                .narwhal_config
                .network_admin_server
                .worker_network_admin_server_base_port = worker_network_admin_server_base_port;
        }
    }
    assert_yaml_snapshot!(network_config, {
        ".genesis" => "[fake genesis]",
        ".validator_configs[].genesis.genesis" => "[fake genesis]",
    });
}
