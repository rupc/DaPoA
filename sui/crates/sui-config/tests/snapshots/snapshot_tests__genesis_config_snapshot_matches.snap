---
source: crates/sui-config/tests/snapshot_tests.rs
expression: genesis_config
---
validator_config_info: ~
parameters:
  timestamp_ms: 0
  protocol_version: 1
  allow_insertion_of_extra_objects: true
  initial_sui_custody_account_address: "0x0000000000000000000000000000000000000000000000000000000000000000"
committee_size: 4
grpc_load_shed: ~
grpc_concurrency_limit: 20000000000
accounts:
  - address: 33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b
    gas_objects:
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
    gas_object_ranges: []
  - address: 33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b
    gas_objects:
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
    gas_object_ranges: []
  - address: 33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b
    gas_objects:
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
    gas_object_ranges: []
  - address: 33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b
    gas_objects:
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
    gas_object_ranges: []
  - address: 33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b
    gas_objects:
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
      - object_id: "0x33d3e552666edc4f048c74e0d2776d0b18171e31c0353e4929e73ba846af602b"
        gas_value: 100000000000000
    gas_object_ranges: []

